api/metrics-api-jaxrs/src/main/java/org/hawkular/metrics/api/jaxrs/handler/observer/MetricObserver.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright 2014-2017 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.hawkular.metrics.api.jaxrs.handler.observer;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import org.hawkular.metrics.model.ApiError;
+import org.hawkular.metrics.model.Metric;
+import org.jboss.logging.Logger;
+
+import com.fasterxml.jackson.core.JsonEncoding;
+import com.fasterxml.jackson.core.JsonGenerator;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.base.Throwables;
+
+import rx.Subscriber;
+
+/**
+ * @author jsanda
+ */
+// TODO Create a base class for MetricObserver and NamedDataPointObserver
+public class MetricObserver<T> extends Subscriber<Metric<T>> {
+
+    private static Logger logger = Logger.getLogger(MetricObserver.class);
+
+    @FunctionalInterface
+    private interface WriteValue<T> {
+        void call(Metric<T> dataPoint) throws IOException;
+    }
+
+    private final HttpServletRequest request;
+    private final HttpServletResponse response;
+    private final JsonGenerator generator;
+    private final ObjectMapper mapper;
+    private AtomicInteger count;
+    private final ByteArrayOutputStream jsonOutputStream;
+
+    private volatile Metric<T> current;
+
+    public MetricObserver(HttpServletRequest request, HttpServletResponse response, ObjectMapper mapper) {
+        this.request = request;
+        this.response = response;
+        this.mapper = mapper;
+        count = new AtomicInteger();
+        jsonOutputStream = new ByteArrayOutputStream();
+        try {
+            generator = mapper.getFactory().createGenerator(response.getOutputStream(), JsonEncoding.UTF8);
+            generator.writeStartArray();
```

Comment:
Shouldn't this happen at the onNext (when the count.incrementAndGet() == 1) ? Now we write to this generator even if we're not going to write anything else to it (only 204 code).